tensorflow_probability/python/mcmc/internal/util.py

```diff
@@ -332,8 +332,10 @@ def smart_for_loop(loop_num_iter, body_fn, initial_loop_vars,
   """
   with tf.name_scope(name or 'smart_for_loop'):
     loop_num_iter_ = tf.get_static_value(loop_num_iter)
-    if (loop_num_iter_ is None or tf.executing_eagerly() or
-        control_flow_util.GraphOrParentsInXlaContext(
+    if (loop_num_iter_ is None
+        or tf.executing_eagerly()
+        or loop_num_iter_ > 1
```

Comment:
I think this 1 is probably too conservative. I've seen numbers <100 to be good compromises, e.g. https://github.com/google/jax/commit/0a45fc3b9bf8ae53b371d0e1175ff37127a6bc34. Maybe make it configurable? Have you tested with anything other than 1?

Reply:
Configurability is reasonable.  That said, I am pretty strongly in favor of being very conservative here by default, because the function being unrolled is unknown.  If I understand your example correctly, the loop they are unrolling is just a multiply-add, where unrolling to 64 iterations makes sense.  But if the loop body is an HMC transition on the LogGaussianCoxProcess model, which is where this came up for us, then unrolling even 2 of them has a decent chance of being counterproductive (but I am not aware of experiments to that effect).

Reply:
Yeah, that makes perfect sense. Was that still true even after tf.function decorating? I'm be fine with configurability + default of 1 loop.

Reply:
I don't know.  We didn't conduct a formal study, except to note that going from 4 copies to 1 (by tf.function decorating the density) took 50% = 1.5s off the graph construction time.  That's not perfectly comparable, because the issue here is about whether to replicate (HMC kernel + 4 * density) num_steps times or once, but in any case, I think we are in agreement.

Reply:
@rupei Clarifying: consensus here is to add an optional argument to `smart_for_loop` named something like `unroll_threshold=1`, and use that instead of the hardcoded 1 limit on unrolls.